crypto/asn1/a_time.c

```diff
@@ -561,3 +561,61 @@ int ASN1_TIME_compare(const ASN1_TIME *a, const ASN1_TIME *b)
         return -1;
     return 0;
 }
+
+ASN1_TIME *asn1_string_to_ASN1_TIME(char *asn1_string)
+{
+    size_t len;
+    ASN1_TIME *tmps = NULL;
+    char *p;
+
+    len = strlen(asn1_string) + 1;
+    tmps = ASN1_STRING_new();
+    if (tmps == NULL)
+        return NULL;
+
+    if (!ASN1_STRING_set(tmps, NULL, len)) {
+        ASN1_STRING_free(tmps);
+        return NULL;
+    }
+
+    if (strlen(asn1_string) == 13)
+        tmps->type = V_ASN1_UTCTIME;
+    else
+        tmps->type = V_ASN1_GENERALIZEDTIME;
+    p = (char*)tmps->data;
+
+    tmps->length = BIO_snprintf(p, len, "%s", asn1_string);
+
+    return tmps;
+}
+
+time_t *asn1_string_to_time_t(char *asn1_string)
+{
+    ASN1_TIME *testdate_asn1 = NULL;
+    struct tm *testdate_tm = NULL;
+    time_t *testdatelocal = NULL;
+    time_t *testdateutc = NULL;
+
+    testdate_asn1 = asn1_string_to_ASN1_TIME(asn1_string);
+    if (testdate_asn1 == NULL)
+        return NULL;
+
+    testdate_tm = malloc(sizeof(*testdate_tm));
+
+    if (!(ASN1_TIME_to_tm(testdate_asn1, testdate_tm))) {
+        free(testdate_tm);
+        ASN1_STRING_free(testdate_asn1);
+        return NULL;
+    }
+
+    testdatelocal = malloc(sizeof(time_t));
+    *testdatelocal = mktime(testdate_tm);
+    free(testdate_tm);
+
+    testdateutc = malloc(sizeof(time_t));
+    *testdateutc = *testdatelocal - timezone;
```

Comment:
Doesnt timezone not do daylight savings..

Sigh.. Playing with time should not be this hard (but it appears to be).
https://stackoverflow.com/questions/32424125/c-code-to-get-local-time-offset-in-minutes-relative-to-utc

Reply:
I wrote some tests for asn1_string_to_time_t just a second before switching to DST in 2021 in Central Europe and this seems to work locally. I didn't find a way to specify a timezone in ASN.1 string, only UTC and difference to UTC. Is there something I didn't find?
If there is a more elegant solution to this, we can change the function and we would see regression errors. I had a look at the stackoverflow link but I'm not sure there are better solutions. The original problem is "mktime" as it converts to localtime. I didn't find a standard function to convert struct tm to time_t resulting in UTC.